Code/GraphMol/MolTransforms/MolTransforms.cpp

```diff
@@ -156,108 +204,106 @@ bool computePrincipalAxesAndMoments(const RDKit::Conformer &conf,
+  if (res && !weights) {
+    conf.getOwningMol().setProp(axesPropName, axes, true);
+    conf.getOwningMol().setProp(momentsPropName, moments, true);
   }
-  // std::cerr<<"  origin: "<<origin<<" "<<wSum<<std::endl;
-  origin /= wSum;
-
-  double sumXX, sumXY, sumXZ, sumYY, sumYZ, sumZZ;
-  computeCovarianceTerms(conf, origin, sumXX, sumXY, sumXZ, sumYY, sumYZ, sumZZ,
-                         true, ignoreHs, weights);
+  return res;
+}
 
-  Eigen::Matrix3d mat;
-  mat << sumXX, sumXY, sumXZ, sumXY, sumYY, sumYZ, sumXZ, sumYZ, sumZZ;
-  // std::cerr<<"  matrix: "<<mat<<std::endl;
-  Eigen::SelfAdjointEigenSolver<Eigen::Matrix3d> eigensolver(mat);
-  if (eigensolver.info() != Eigen::Success) {
-    BOOST_LOG(rdErrorLog) << "eigenvalue calculation did not converge"
-                          << std::endl;
-    return false;
+RDGeom::Transform3D *computeCanonicalTransform(const Conformer &conf,
+                                               const RDGeom::Point3D *center,
+                                               bool normalizeCovar,
+                                               bool ignoreHs) {
+  constexpr unsigned int DIM = 3;
+  RDGeom::Point3D origin;
+  if (!center) {
+    origin = computeCentroid(conf, ignoreHs);
+  } else {
+    origin = (*center);
   }
+  unsigned int nAtms = conf.getNumAtoms();
+  auto *trans = new RDGeom::Transform3D;
+  trans->setToIdentity();
 
-  axes = eigensolver.eigenvectors();
-  moments = eigensolver.eigenvalues();
-  if (!weights) {
-    conf.getOwningMol().setProp(axesPropName, axes, true);
-    conf.getOwningMol().setProp(momentsPropName, moments, true);
+  // if we have a single atom system we don't need to do anyhting setting
+  // translation is sufficient
+  if (nAtms > 1) {
+    Eigen::Matrix3d eigVecs;
+    Eigen::Vector3d eigVals;
+    if (getEigenValEigenVectFromCovMat(conf, eigVecs, eigVals, origin, ignoreHs,
+                                       normalizeCovar, nullptr)) {
+      std::vector<std::pair<unsigned int, double>> eigValsSorted;
+      CHECK_INVARIANT(eigVals.size() == DIM, "less eigenvalues than expected");
+      auto eigVecCoeffSums = eigVecs.colwise().sum();
+      eigValsSorted.reserve(DIM);
+      for (unsigned int i = 0; i < DIM; ++i) {
+        eigValsSorted.emplace_back(i, eigVals(i));
+      }
+      std::sort(eigValsSorted.begin(), eigValsSorted.end(),
+                [](const std::pair<unsigned int, double> &a,
+                   const std::pair<unsigned int, double> &b) {
+                  return (a.second > b.second);
+                });
+      for (unsigned int col = 0; col < DIM; ++col) {
+        unsigned int colSorted = eigValsSorted.at(col).first;
+        double sign = (eigVecCoeffSums(colSorted) > 0.0) ? 1.0 : -1.0;
+        for (unsigned int row = 0; row < DIM; ++row) {
+          trans->setVal(col, row, sign * eigVecs(row, colSorted));
+        }
+      }
+    }
   }
-  return true;
-}
-#endif
+  origin *= -1.0;
```

Comment:
I guess this is more efficient, and thus potentially worth doing, but I find it much simpler to set up a transformation matrix for the translation to the origin and then to right multiply that onto the transformation matrix for the rotation.

Reply:
I thought the same when I saw that code (which pre-existed my PR - I have only copied it across). But it's indeed a bit more efficient as it involves only in-place transformation of a vector rather than creation of another `Transform3D` object followed by matrix multiplication. This is why I left it as it was.